File: src/main/scala/com/tort/mudai/event/KillTrigger.java
package com.tort.mudai.event;

import com.tort.mudai.Handler;
import com.tort.mudai.task.AbstractTask;
import com.tort.mudai.task.EventDistributor;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class KillTrigger implements EventTrigger {
    private final EventDistributor _eventDistributor;
    private static final Pattern PATTERN = PatternUtil.compile(".*\u001B\\[0\\;37m([^\n]*) (?:мертв|мертва|мертвы), (?:его|ее) душа медленно подымается в небеса.\r?\n" +
            "Ваш опыт повысился на ([1234567890]*) (?:очка|очков?)\\.\r?\n.*");

    public KillTrigger(EventDistributor eventDistributor) {
        _eventDistributor = eventDistributor;
    }

    @Override
    public Event fireEvent(String text) {
        final Matcher matcher = PATTERN.matcher(text);
        matcher.find();
        final String target = matcher.group(1);

        _eventDistributor.invoke(new Handler(){
            @Override
            public void handle(AbstractTask task) {
                task.kill(target);
            }
        });

        return null;
    }

    @Override
    public boolean matches(String text) {
        return PATTERN.matcher(text).matches();
    }
}

File: src/main/scala/com/tort/mudai/event/KillTrigger.scala
package com.tort.mudai.event

import com.tort.mudai.mapper.Mob._
import scalaz.@@

class KillTrigger extends EventTrigger[KillEvent] {
  val Pattern = ("""(?ms).*\n(?:\u001B\[\d\;\dm)?([^\n]*) (?:мертв[аыо]?), (?:его|ее) душа медленно подымается в небеса\..*""" +
            """Ваш опыт повысился на ([1234567890]*) (?:очк[оа]в?)\..*Кровушка стынет в жилах от предсмертного крика ([^\n]*)\..*"""
    ).r

  def matches(text: String) = Pattern.findFirstIn(text).isDefined

  def fireEvent(text: String) = {
    val Pattern(target, exp, gen) = text
    KillEvent(shortName(target), exp.toInt, genitive(gen), magic = false)
  }
}


class KillMagicMobTrigger extends EventTrigger[KillEvent] {
  val Pattern = ("""(?ms).*\n(?:\u001B\[\d\;\dm)?([^\n]*) вспыхнул[иао]? и рассыпал[аио]?с[ья] в прах\..*""" +
            """Ваш опыт повысился на ([1234567890]*) (?:очк[оа]в?)\..*Кровушка стынет в жилах от предсмертного крика ([^\n]*)\..*"""
    ).r

  def matches(text: String) = Pattern.findFirstIn(text).isDefined

  def fireEvent(text: String) = {
    val Pattern(target, exp, gen) = text
    KillEvent(shortName(target), exp.toInt, genitive(gen), magic = true)
  }
}

case class KillEvent(target: String @@ ShortName, exp: Int, genitive: String @@ Genitive, magic: Boolean) extends Event
